### PR TITLE
KEP-1967: promote size backed memory volumes to stable

### PR DESCRIPTION
**Pull request by kannon92**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Promote KEP-1967 to stable.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
Promote SizeMemoryBackedVolumes to stable
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kannon92** [comment]
/hold

This is dependent on the KEP outcome.

**Comment by kannon92** [comment]
/hold cancel

KEP was merged so this is ready for review/approval.

File: test/e2e_node/eviction_test.go
Comment (kannon92) [inline test/e2e_node/eviction_test.go:303]: Removed this test because it is actually testing sizeMemoryBackedVolumes with the gate off. The entire test can be removed.

File: test/e2e_node/eviction_test.go
Comment (SergeyKanzhelev) [inline test/e2e_node/eviction_test.go:290]: Do yous still need the `feature.LocalStorageCapacityIsolation`? I think we can switch this to NodeConformance

File: pkg/volume/emptydir/empty_dir_test.go
Comment (SergeyKanzhelev) [inline pkg/volume/emptydir/empty_dir_test.go:22]: Also this needs to be removed: https://github.com/kubernetes/kubernetes/blob/c15581b277e9e42d830898357fdbd92dcf7981c6/pkg/volume/emptydir/empty_dir.go#L115

File: test/e2e_node/eviction_test.go
Comment (kannon92) [reply test/e2e_node/eviction_test.go:290]: Maybe the feature but this test is in the eviction suite so it’s really long running and quite disruptive. I don’t think we should move this test into conformance.

I don’t think this is related to this PR though.

File: pkg/volume/emptydir/empty_dir_test.go
Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:22]: I thought in two releases I can remove the feature gate.

Comment (SergeyKanzhelev) [reply pkg/volume/emptydir/empty_dir_test.go:22]: It is locked to true. So you don't need this check. Ideally, removal of feature gate do not touch any code besides the definition of a feature gate itself

File: test/e2e_node/eviction_test.go
Comment (SergeyKanzhelev) [reply test/e2e_node/eviction_test.go:290]: We either need to explain what exactly is special about the test environment it needs to run in here: https://github.com/kubernetes/kubernetes/blob/2f7df335ad2d2ec8c3e876cbd2f4e5b3d3e17931/test/e2e/feature/feature.go#L208-L209 or remove the feature altogether.

It is already marked as Serial so it will not run with non-serial NodeConformance. Should it be running with eviction tests?

Comment (kannon92) [reply test/e2e_node/eviction_test.go:290]: You are commented on a feature that is not related to this KEP.

https://kubernetes.io/blog/2022/09/19/local-storage-capacity-isolation-ga/

This was already promoted to GA. I mainly changed this because it was using the feature gate.

> It is already marked as Serial so it will not run with non-serial NodeConformance. Should it be running with eviction tests?

Yes. It is using the eviction framework.

File: pkg/volume/emptydir/empty_dir_test.go
Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:22]: Removed.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

**Comment by kannon92** [comment]
/assign @saad-ali @thockin 

We still need an approval on the emptyDir code.

File: pkg/features/kube_features.go
Comment (thockin) [inline pkg/features/kube_features.go:748]: 35 - the new version-compat requires 3-deep memory, so 34 has to be able to mimic 33, 32, 31.  In 31 this was beta.

File: pkg/volume/emptydir/empty_dir.go
Comment (thockin) [inline pkg/volume/emptydir/empty_dir.go:115]: This all needs to stay because of version-compat.  Right @jpbetz  ?

https://github.com/kubernetes/community/issues/8082

**Comment by thockin** [comment]
The biggest concerns I have are around accounting, especially how it intersects with container memory requests and limits.

If a container opens and writes files, wo which cgroup is that charged?  The container or the pod?

What happens if the container crashes and is restarted?

What happens if the container has requests: {memory: 1Gi}, limits: {memory: 1Gi}  and the volume has a size-limit of 2Gi ?  If I write to the volume, can I cause the pod to fail to start?

Do we have e2es that verify the accounting?

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:748]: where can I read how this version compat works? We are locking the feature gate to default so it is not possible to disable the fearure gate. What would version compat do?

Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:748]: Also, we will need to preserve all the tests too? Like ones that disable the feature gate?

I thought compat mode was only for api server

**Comment by kannon92** [comment]
> If a container opens and writes files, wo which cgroup is that charged? The container or the pod?

The container cgroup.

> What happens if the container crashes and is restarted?
I don't really understand how this is related to this feature. What exactly are you concerned about? This feature uses cgroup of the container to set a max memory limit. I don't see how that would be impacted by restarts.

> What happens if the container has requests: {memory: 1Gi}, limits: {memory: 1Gi} and the volume has a size-limit of 2Gi ? If I write to the volume, can I cause the pod to fail to start?

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: example
  labels:
    app: test-pd
spec:
  securityContext:
    seccompProfile:
      type: RuntimeDefault
  containers:
  - image: busybox
    command:
      - /bin/sh
      - -c
      - |
         sleep infinity
    name: test-pd
    resources:
      limits:
        memory: 2Gi
    volumeMounts:
    - mountPath: /dev/shm
      name: dshm
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
  volumes:
  - name: dshm
    emptyDir:
      medium: Memory
      sizeLimit: 4Gi
```
In this case, /dev/shm would have a limit of 2Gi rather than 4Gi. It seems to take the memory limit of the pod over sizeLimits. 

Do we have e2es that verify the accounting?

Yes. We have a test [here](https://github.com/kubernetes/kubernetes/blob/468676cb2f5a84647823e2bd6b5133ac859b2f6f/test/e2e_node/eviction_test.go#L337).

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go]: I updated this to say remove for 1.35.

File: pkg/volume/emptydir/empty_dir.go
Comment (kannon92) [reply pkg/volume/emptydir/empty_dir.go:115]: There seems to be some confusion here so I'll wait for consensus. It makes sense to remove this to me as this is effectively dead code. (GA feature gates will never go to false). I think we should not remove the true part of the feature gate but I will wait on consensus.

File: pkg/features/kube_features.go
Comment (thockin) [reply pkg/features/kube_features.go]: @jpbetz and @Jefftree were discussion how best to update docs.  The compat is IIUC, fotr the whole system - APIs and controllers.

**Comment by thockin** [comment]
>> What happens if the container crashes and is restarted?
>
> I don't really understand how this is related to this feature. What exactly are you concerned about? This feature uses cgroup of the container to set a max memory limit. I don't see how that would be impacted by restarts.

IIUC A write to tmps creates dirty anonymous pages (can't be flushed to disk for reclaim) attributed to the container's cgroup.  If that cgroup were to get torn down for any reason, those pages are still anonymous and dirty, so they get accounted by the next-parent cgroup (I think?).  If this happens, you can imagine a situation where the container tries to restart and immediately OOMs because those tmpfs files are accounted to the pod.

What I am asking is whether we have thought through those sorts of failure modes.  Do we ever tear down the container cgroup?  Are we confident that the accounting stays with the container?  If you all say we are satisfied with the testing of that, I'm satisfied, but I wanted to call it out as an area where I know there are traps :)

File: pkg/volume/emptydir/empty_dir.go
Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir.go:115]: Yes, please keep the feature gate code. Yes it used to be dead code but with the introduction of [Compatibility Versions](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md), kubernetes is able to emulate a previous version that could have the feature gate disabled. We are in the process of updating the documentation for all this, thank you for catching this Tim!

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go]: https://github.com/kubernetes/kubernetes/pull/126981#discussion_r1799432320

File: pkg/volume/emptydir/empty_dir.go
Comment (kannon92) [reply pkg/volume/emptydir/empty_dir.go:115]: Okay. I will add this back. So when we remove the feature gate in n+3, we can remove the dead code.

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir.go]: removed.

Comment (thockin) [reply pkg/volume/emptydir/empty_dir.go]: Yeah, instead if it being "set GA and delete gated code; wait +2; remove gate" it's now "set GA; wait +3; delete gated code and remove gate"

File: pkg/volume/emptydir/empty_dir_test.go
Comment (thockin) [inline pkg/volume/emptydir/empty_dir_test.go:966]: @Jefftree how do we feel about removing gated tests at GA?  Even though the gat is "locked", there is still code under the gate and ultimately it's just a variable, right? Should we keep the tests until N+3?

File: pkg/volume/emptydir/empty_dir.go
Comment (SergeyKanzhelev) [reply pkg/volume/emptydir/empty_dir.go]: > We are in the process of updating the documentation for all this

I am a bit confused how compatibility mode can be enabled and what the "LockToDefault" would mean. Documentation would be awesome to have. I am mostly interested on how we will guarantee functionality without the feature gate and what testing is run (or needs to run) to ensure functionality.

File: pkg/volume/emptydir/empty_dir_test.go
Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: I think we have to remove it. If a test tries to change a feature gate I think it is a panic for GA flags.

Comment (SergeyKanzhelev) [reply pkg/volume/emptydir/empty_dir_test.go:966]: yeah, it may indicate that we either need to rewrite this test to enable compatibility mode instead. But we need to see what kind of compatibility guarantees we are documenting.

Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir_test.go:966]: You can keep the test by adding a line `featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))` before setting the feature gate via `featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.SizeMemoryBackedVolumes, testCase.featureGateEnabled)`. An example can be seen here: https://github.com/kubernetes/kubernetes/pull/127348/files#diff-92f41de420a097c0419a3cb53cf176a459a742c39f3fa482ab48b1c2fa6fd084R6730.

The disablement test will be locked to emulation version 1.31 that k8s versions 1.32-1.34 will support.

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: I don't think this works here. For E2E node, we set the feature gate on the kubelet config.

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: sorry, this would be tricky for https://github.com/kubernetes/kubernetes/pull/126981/files#diff-587ea2030f2b92ebe8061f12bf42ae3027a51a9c43ff16ae8a77f7a9a1aa167c.

I think for the unit tests I can add this.

File: pkg/volume/emptydir/empty_dir.go
Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir.go]: Thanks for the feedback and this is definitely something that we will clarify in the documentation. 

Per the [Compatibility Version KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-architecture/4330-compatibility-versions/README.md#--emulation-version), a k8s cluster at a particular binary version can emulate a previous version up to three releases before (min version starts at 1.31). Features/APIs/storage-versions/validation will emulate the behavior at the target emulation version. So a 1.32 binary version cluster emulating 1.31 will have `SizeMemoryBackedVolumes` in beta and disableable. LockToDefault is only applied to emulation version 1.32+ (hence why feature promotions are all versioned now).

Comment (SergeyKanzhelev) [reply pkg/volume/emptydir/empty_dir.go]: Wait, KEP explicitly says: "Adding --emulation-version to kubelet is out of scope for this enhancement."

I was always under impression that for nodes we are saying just to install a previous kubelet version. Has this changed?

Comment (SergeyKanzhelev) [reply pkg/volume/emptydir/empty_dir.go]: xref to another KEP to update on the decision here: https://github.com/kubernetes/kubernetes/pull/128046#discussion_r1799759192

Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir.go]: > I was always under impression that for nodes we are saying just to install a previous kubelet version. Has this changed?

Correct, kubelet is unaffected by compatibility versions.

File: pkg/volume/emptydir/empty_dir_test.go
Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir_test.go:966]: deletion in e2e node should be fine because compatibility version is not supported in kubelet. The comment is referring to unit and integration tests.

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: So should we be changing all unit tests that trigger features gates to use `featuregatetesting.SetFeatureGateEmulationVersionDuringTest(t, utilfeature.DefaultFeatureGate, version.MustParse("1.31"))`?

Or something like that?

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: I'm not sure what is required from me for this and what we should enforce on all PRs for the next few weeks.

Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir_test.go:966]: Sorry I'm trying to understand this a bit better, is this used at all by control plane components or is it a kubelet only feature? Kubelet only features require no feature gate changes since compatibility version is not supported so if that's the case then this PR is lgtm for feature gates.

Keeping the tests described in https://github.com/kubernetes/kubernetes/pull/126981/files#r1801852232 is only necessary for control plane components. I see the volume package is imported by kcm though am having a bit of trouble understanding the full graph. 

Thanks!

cc @thockin @jpbetz

Comment (kannon92) [reply pkg/volume/emptydir/empty_dir_test.go:966]: The KEP states that Kubelet is the only component where this feature gate is used. Volumes confuse me a bit also but I think it is only kubelet. There isn't any API changes for this feature and it only changes some basic behavior with emptyDir.

Comment (Jefftree) [reply pkg/volume/emptydir/empty_dir_test.go:966]: Okay, feature gate changes lgtm.

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:748]: can you please add a comment saying that this FG is only used in kubelet and not needed for emulated version so in couple releases it will be easier to understand that we can just remove it

Comment (kannon92) [reply pkg/features/kube_features.go:748]: I will do this tonight or tomorrow.

**Comment by kannon92** [comment]
@thockin 
As far as I can tell, we are not using pod c groups to track memory. I looked into the pod cgroup of a pod to see if we report memory cgroup settings for the containers. I don’t think we use the chorus for tracking (memory.current is set to 0 even though container memory.current is nonzero). Memory.max for pods was set to max.

So I don’t think we are tracking anything in the pod c group.

File: pkg/features/kube_features.go
Comment (kannon92) [reply pkg/features/kube_features.go]: done.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm
/approve

I think we resolved all the questions here.

**Comment by kannon92** [comment]
@thockin I asked @ndixita her thoughts on the pod tracking. This is not a concern because we haven't been using placing tracking information into the pod cgroup.

> What I am asking is whether we have thought through those sorts of failure modes. Do we ever tear down the container cgroup? Are we confident that the accounting stays with the container? If you all say we are satisfied with the testing of that, I'm satisfied, but I wanted to call it out as an area where I know there are traps :)

I took this item because of its long outstanding status as a beta feature (quote is from you).
> This gate has been Beta since 1.22 - is there any reason not to just set it to GA? Seems like something any contributor can do to score some brownie points, and since it has been beta so long, I doubt we can change any behavior, even if we want to?

I am paraphrasing you here but this feature has been in beta on since 1.22. At this stage, I think any new issue related to emptyDir and tmpfs should be considered separate from this issue. This feature was about setting a size limit to node allocatable or pod limit. It didn't change anything around how the cgroups are tracked from container restarts. I think the feature works as intended and we have e2e tests that verify the limits.

**Comment by kannon92** [comment]
@thockin 
> What happens if the container has requests: {memory: 1Gi}, limits: {memory: 1Gi} and the volume has a size-limit of 2Gi ? If I write to the volume, can I cause the pod to fail to start?

https://github.com/kubernetes/kubernetes/issues/128339

Yes. It looks like if a container hits an OOM limit with tmpfs than it will be stuck in an error state. I don't think this feature introduced this though.

**Comment by SergeyKanzhelev** [comment]
Even if we want to say that memory backed volume must have limit set and it must be lower than container limit, we will not be able to enforce it on API side without breaking existing Pods.

So the only option would be for kubelet to override the memory-backed volume size limit to the container's limit (when defined). This may help. But it also will not guarantee we out of the "OOMLoopBackoff" as code bootstrap can be quite memory-heavy.

I would see this as to be outside of the KEP scope, but it can be a good future enhancement. @thockin do you agree?

**Review by thockin — COMMENTED** [review]
Thanks for doing the homework.  I know how baroque some of the corner-cases are.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**
